### ISSUE TITLE
Complex derivative for conjugate

### ISSUE THREAD

**Issue by anjali411** [google/jax#3293]
The obtained result seems to be inconsistent with the formula stated in this [comment](https://github.com/google/jax/issues/979#issuecomment-508864516) 
For conjugate function, the matrix equals [[1, 0], [0, -1]], so the gradient should be 1 -1j based on the complex convention stated [here](https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#Complex-numbers-and-differentiation) and [here](https://github.com/HIPS/autograd/blob/master/docs/tutorial.md#complex-numbers), unless I am missing something.

```
>>> def conj(x):
...     return x.conj().sum()
...
>>> conj_grad=grad(conj, holomorphic = True)
>>> conj_grad(np_jax.array([1+2.j,2.+3j,3.+4j]))
DeviceArray([1.-0.j, 1.-0.j, 1.-0.j], dtype=complex64)
```

**Comment by anjali411** [comment]
@mattjj did you get a chance to look at the issue? also how does `check_grads` work for complex, i.e. which direction is looked into to calculate the numerical gradient for complex functions?

**Comment by mattjj** [comment]
Sorry for the delay!

I think writing `grad(conj, holomorphic = True)` might be problematic because complex conjugation is not holomorphic; it doesn't preserve orientation for example, and moreover taking u(x,y)=x and v(x,y)=-y we have ∂_0 u = 1 and ∂_1 v = -1.

Even though `grad` is nominally about computing a derivative (i.e. Jacobian), the way it works is by computing a VJP, namely `grad(f)(x)` is computed as `vjp(f, x)[1](1.)`. If f is an R->R function then this reveals the Jacobian, since we're basically doing the Jacobain (transpose) times 1. Also, if f is holomorphic, meaning its derivative can be represented by a single complex number, then this _still_ reveals the full Jacobian. But if f is _not_ holomorphic then the correspondence between Jacobian and vjp-of-one no longer holds.

When you write `holomorphic=True`, you're promising the function is really holomorphic so that we can proceed with the usual vjp-of-one calculation. For non-holomorphic functions, you can still compute the Jacobian or whatever else you want; you just need to write your own function in terms of `vjp` (and/or `jvp`) and can't use the `grad` convenience wrapper.

WDYT? Am I missing anything?

**Comment by mattjj** [comment]
> also how does check_grads work for complex, i.e. which direction is looked into to calculate the numerical gradient for complex functions?

[Here's the implementation of `check_grads`](https://github.com/google/jax/blob/05fde0f1709783610e3ccaaaf8ba682a9691c750/jax/test_util.py#L199-L270); it's quite short! It uses the [`rand_like` function](https://github.com/google/jax/blob/05fde0f1709783610e3ccaaaf8ba682a9691c750/jax/test_util.py#L172-L179) to generate random tangent or cotangent vectors to push forward or pull back with `jvp` or `vjp`, respectively, based on the primal input/output types. That way, if a function has complex inputs or complex outputs (i.e. if the inputs/outputs indicate that it's R->C or C->R or C->C), we generate perturbation vectors/covectors appropriately.

In the context of this issue, recall that `jvp` and `vjp` don't have to care about holomorphicity because they aren't trying to identify a derivative (i.e. Jacobian or gradient) with the result of any calculation; it's always clear what `jvp` and `vjp` mean!

I think we covered the bases of this issue, so I'm going to close it. Please reopen or open a new one as you have more questions!

**Comment by anjali411** [comment]
@mattjj Thanks for your response. that makes sense. However I am not entirely sure about how the VJP will be written for C->R and R->C functions like:
`x + iy -> [x, y]` and `[x, y] -> x+iy`

**Comment by mattjj** [comment]
Check out [this recently-revised section in the autodiff cookbook](https://colab.research.google.com/github/google/jax/blob/master/docs/notebooks/autodiff_cookbook.ipynb#scrollTo=3fPWLrxK8bgD), and also [this ongoing discussion](https://github.com/google/jax/issues/3402). The short answer is to think in terms of R^2->R and R->R^2 maps, respectively, and R-linearity.

**Comment by anjali411** [comment]
The first function `f(x+iy) = [x, y]` is C->R^2 so do you mean writing it's jacobian as if it were R^2 -> R^2. JVP:
```
[gx   -gy] x [[1, 0], [0, 1]] [1, -i]^T
```

JVP for f([x, y]) = x+iy (R^2 -> C) will be:
```
[gx   -gy] x [[1, 0], [0, 1]] [1, -i]^T
```
If we follow this the JVP for `f(x+iy) = [x, y] and f([x, y]) = x+iy` will turn out to be equal which looks weird (unless I am missing something?)

**Comment by mattjj** [comment]
I was just referring to R->C and C->R functions, not the particular examples you wrote.

I think JAX's rules are unambiguous and useful; take a look at the notebook if you haven't already. Otherwise maybe we can follow up in the other thread!
